### PR TITLE
Upgrade Terraform Version Used from 0.12 to 0.13

### LINKED ISSUE #297: Update Terraform Code for Terraform 0.13
# 🚀 Feature Proposal

With [Terraform 0.14 having released](https://github.com/hashicorp/terraform/releases/tag/v0.14.0), this project needs to be updated to work with Terraform 0.13.

## Motivation

Terraform 0.13 added functionality to make modules "first-class citizens" of Terraform development. The [release notes](https://github.com/hashicorp/terraform/releases/tag/v0.13.0) have more specific information. Given the reliance on modules in this project, this may prove beneficial. Additionally it is good if we aren't too far behind in Terraform versions so we see regular updates.

### LINKED ISSUE #326: Update to use the 3.x AWS Provider
# 🚀 Feature Proposal

The AWS provider used in this project should be updated to use the `3.x` major version as we use in our other Terraform projects at this point.

## Motivation

The `3.x` version has stabilized and our COOL related projects have already migrated to using this AWS provider version. This project is unable to take advantage of new features because it is still pinned against the `2.x` version.

## Pitch

We should be getting access to new functionality as well as improved functionality and stability on the currently developed version of the AWS provider.

### LINKED ISSUE #327: Use `default_tags` in Providers
# 🚀 Feature Proposal

We should switch to using the `default_tags` block in providers to manage common tags. The version of the AWS provider this requires means this relies on #326 being resolved before it can be implemented.

## Motivation

This would follow what was done in https://github.com/cisagov/skeleton-tf-module/pull/49 and is implemented (or in the process of being implemented) for other Terraform projects now.

## Pitch

Although this project descends from [cisagov/skeleton-generic](https://github.com/cisagov/skeleton-generic) we should still try to incorporate changes made in [cisagov/skeleton-tf-module](https://github.com/cisagov/skeleton-tf-module) and [cisagov/skeleton-packer](https://github.com/cisagov/skeleton-packer) whenever possible.

### PR DESCRIPTION
**Pull request by mcdonnnj**
#  #

## 🗣 Description ##

This PR updates the Terraform version that this project pins against from Terraform 0.12 to 0.13 and makes corresponding changes to the Terraform configuration.

## 💭 Motivation and context ##

This mirrors work in https://github.com/cisagov/cool-system/issues/94 to bring our entire org to Terraform 0.13 at the same time as bringing the COOL configurations (which are the vast majority of our Terraform code) to Terraform 0.13. The initial changes were done using `terraform 0.13upgrade` with a Terraform 0.13 binary. However, due to changes in Terraform 0.13 to bring full functionality to modules this PR also had to dismantle the previous approach of dynamic modules for Nessus and nmap instances and makes them regular modules in the Terraform configuration.

## ⚙ Process ##

I first worked on a clean `terraform0.12 apply` with the existing Terraform 0.12 state for each Terraform configuration in this repository (`terraform_egress_pub/`, `terraform_route_53/`, `terraform_nessus_only/`, `terraform/`) using the `develop` branch. Once that was ensured I switched to this branch, ran `terraform0.13 init -reconfigure`, and then attempted a `terraform0.13 apply`. If there were any issues I resolved them and proceeded with the `apply` until it was clean.

### `terraform_egress_pub/` ###

This had no issues with the existing Terraform 0.12 state and was fine for a Terraform 0.13 `apply` with the existing code.

### `terraform_route_53/` ###

When attempting to perform a `terraform0.12 apply` the resource

https://github.com/cisagov/cyhy_amis/blob/f4551046bb205dc5dedf517745af45731001ca97/terraform_route_53/cyber.dhs.gov.tf#L122-L128

needed to be created. This should exist as part of the configuration described in [ciagov/cool-dns-cyber.dhs.gov](https://github.com/cisagov/cool-dns-cyber.dhs.gov/blob/6a4050b4480f1985fa20623f836bff35dec72783/route53_dmarc_app.tf#L38-L46), but for some reason did not. After a team discussion we decided to let it get created to move forward. Once this apply was completed there were no issues with a successful Terraform 0.13 `apply`.

### `terraform_nessus_only/` ###

We have determined that this configuration will be completely spun-down and something equivalent will be spun up in the COOL. This configuration **remains** Terraform 0.12 and will have a `terraform0.12 destroy` once the migration occurs as opposed to updating the state to Terraform 0.13.

### `terraform/` ###

Setup for this Terraform configuration's migration included adding

```hcl
lifecycle {
  ignore_changes = [user_data, user_data_base64]
}
```

to the EC2 configurations to prevent changes unrelated to the Terraform 0.13 migration from forcing an instance redeploy.

When attempting to perform a `terraform0.12 apply` there were a number of resources that Terraform deemed necessary to create as well as changed from switching to `v3` of the AWS provider as that change had not been applied. I resolved to fix the missing resources using `v2` of the AWS provider before switching to `v3` and applying those changes.

The resources Terraform wished to create were:

https://github.com/cisagov/cyhy_amis/blob/f4551046bb205dc5dedf517745af45731001ca97/terraform/moe_user_read.tf#L1-L13

https://github.com/cisagov/cyhy_amis/blob/f4551046bb205dc5dedf517745af45731001ca97/terraform/moe_user_read.tf#L15-L45

https://github.com/cisagov/cyhy_amis/blob/f4551046bb205dc5dedf517745af45731001ca97/terraform/moe_user_read.tf#L48-L51

However the first two resources already exist but are not part of the Terraform state and a managed policy with the same permissions as the third resource exists. I approached this by manually importing into the Terraform state the existing resources that mirrored the first two resource definitions and allowing the third to be created as it should not be an issue for a duplicate policy to exist since the permissions are identical.

Once a clean `terraform0.12 apply` could be made I switched to `v3` of the AWS provider and ran a `terraform0.12 apply` which consisted solely of in-place updates to reflect changes in how the provider versions organized information.

When I initially attempted a `terraform0.13 apply` I found that Terraform was unable to interface with the dynamic modules used for the Nessus and nmap Ansible provisioners that were required for Terraform 0.12. I resolved this by switching back to Terraform 0.12 and using the following script to destroy the existing Ansible provisioners

```bash
#!/usr/bin/env bash

set +o nounset
set +o errexit
set +o pipefail

function destroy_items () {
  tf_args=()
  for index in $(seq "$2" "$3"); do
    tf_args+=("-target=module.dyn_$1.module.cyhy_$1_ansible_provisioner_$index")
  done

  terraform apply -var-file="$workspace.tfvars" "${tf_args[@]}"
}

workspace=$(terraform workspace show)

destroy_items "nessus" 0 3
destroy_items "nmap" 0 79
```

which was based on `cyhy_amis/terraform/deploy_new_portscan_instance.sh`/`cyhy_amis/terraform/deploy_new_vulnscan_instance.sh`. Once this was done I was able to perform a `terraform0.13 apply` and once all the provisioners had been either created (dynamic Nessus/nmap) or updated (everything else) it was a clean `apply` after.

## 🧪 Testing ##

I ensured that every configuration passed a `terraform0.13 validate` and that a clean `terraform0.13 apply` occurred per the above process.

## ✅ Checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

### CODE REVIEW

**Comment by dav3r** [comment]
@mcdonnnj - is this PR ready for review?

**Review by jsf9k — APPROVED** [review]

File: terraform/variables.tf [hunk 1]
```diff
@@ -23,6 +29,11 @@ variable "mongo_disks" {
   }
 }
 
+variable "mongo_instance_count" {
```
Comment (dav3r) [inline terraform/variables.tf:32]: This can probably also be given a default value of 1.

**Review by dav3r — APPROVED** [review]
This looks great to my eyes! 🚀 

@mcdonnnj - Thank you so much for all of the work you put in to get this repository updated for Terraform 0.13 - I really appreciate it! 🙌

File: terraform/variables.tf (hunk 1 above)
Comment (mcdonnnj) [reply terraform/variables.tf:32]: Good call! Please see 46576b5 for that default being added.

**Pull request merged**